### DISCUSSION TITLE
Moving file under git lfs from one directory to another within the same repository

### DISCUSSION THREAD

**Discussion by markcoletti** [git-lfs/git-lfs#5035] [Q&A]
I have a file that is under git lfs control that I want to move from one directory to another within the same git repository. However, `git mv foo/bar baz/` works for regular files, but does not bring along the LFS state, so I cannot use that.

I.e., I was thinking along the lines of:
```
git mv foo/biglfs.file bar/
git lfs track *.file
git add bar/.gitattributes
git commit
But I don’t think that properly updates the LFS file pointers, and I think the first statement will actually add biglfs.file to the repo directly, which isn't desirable.
```
I think I have to manually remove the file from under git lfs control, move the file, and then re-add it. But that strikes me as a lot of work, so surely there's an easy and direct way to do this?

**Comment by markcoletti** [comment]
Yep, this is in stackoverflow: https://stackoverflow.com/questions/72354863/moving-file-under-git-lfs-from-one-directory-to-another-within-the-same-reposito

And now I'm here.

**Reply by markcoletti** [reply]
So, I moved the `.gitattributes` to the new directory, but when I do a fresh clone of the repo I cannot use git lfs to checkout/pull that file. It's as if it has had a lobotomy.

**Comment by bk2204** [comment]
Hey,

Can you share with us the contents and location of the `.gitattributes` file before the change?

Typically, one places the `.gitattributes` file at the top of the repository and tracks a general pattern, such as `*.file` (note that you definitely should quote that at the command line, such as `git lfs track "*.file"`).  If you do that, then you can move the file all over the repository without any problems.  If you've done something different in this case, let us know what it is, and we'll try to help.

**Reply by markcoletti** [reply]
Indeed, the file was in the same directory as the large file, in this case an sqlite3 database.  The file just contained this:

`dlfa.db filter=lfs diff=lfs merge=lfs -text`

**Reply by bk2204** [reply]
What you typically want to do in this case is put the `.gitattributes` file at the top of the repository.  There are a handful of reasons why you could something different, but all of those reasons are advanced and should only be done if you know what you're doing.

So in this case, I'd move the `.gitattributes` file to the top level directory first (`git mv .gitattributes "$(git rev-parse --show-toplevel)"`) and then you can move your LFS file.  Even if you've already moved the LFS file, you can just move the `.gitattributes` file, and it should work.

However, if you already have a `.gitattributes` file at the top level, then you'll need to add the line to that top level file.

Now, if you happen to have a file that's also named the same thing elsewhere in your repository, then it will also become an LFS file.  To verify that everything's sorted, you can run `git add --renormalize .` once you've moved the `.gitattributes` entry and then commit and push, which will make sure all your files are properly normalized.

Once you've done that, on a fresh clone, the LFS file should appear as normal.

**Reply by markcoletti** [reply]
I did the following (while you were writing your last response, I was trying things):

1. I moved the `.gitattributes` to top-level directory
2. changed `dlfa.db` to `*.db` manually
3. since that produced no joy, I then did `git lfs untrack "*.db"`
4. verified that now `.gitattributes` at the top-level was empty
5. `git lfs track "*.db"`
6. `git add db/dlfa.db`
7. `git lfs ls-files`
`b63cf9d24d * db/dlfa.db`
8. `git commit -m "added back"`
9. `git push`

Then I verified on github that the file was under LFS control.  Phew!

I now know to do `git track "*foo"` from the top-level directory.  I think part of my reasoning was thinking on a *per file* basis, and not via a *globbing* basis.  I.e., just add an individual file, and that it's ok to do so from the sub-directory in which it resides -- not so!

This was a bit of a painful leveling up experience in git lfs, but there ya go. Thank you for your help, @bk2204 !
